stout/take.h

```diff
@@ -0,0 +1,191 @@
+#include <vector>
+
+#include "stout/eventual.h"
+#include "stout/filter.h"
+#include "stout/stream.h"
+
+namespace stout {
+namespace eventuals {
+namespace detail {
+
+////////////////////////////////////////////////////////////////////////
+
+struct _TakeLastN {
+  template <typename K_, typename Arg_>
+  struct Continuation : public TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, size_t n)
+      : k_(std::move(k)),
+        N_(n) {}
+
+    void Start(TypeErasedStream& stream) {
+      stream_ = &stream;
+      data_ = std::vector<int>(N_);
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      if (i_ == data_.size()) {
+        i_ = 0;
+        filled_ = true;
+      }
+      if (filled_) {
+        for (size_t j = 0; j < data_.size() - 1; ++j) {
+          data_[j] = data_[j + 1];
+        }
```

Comment:
This process of moving everything up a step is `O(N)` copies ([big-O notation](https://www.freecodecamp.org/news/big-o-notation-why-it-matters-and-why-it-doesnt-1674cfa8a23c/)), and since this code will be called `N` times we have an overall `O(N^2)` performance here. We can do better.

The fix is to use a data structure that makes popping-from-front and pushing-onto-back an `O(1)` operation. For example [std::deque](https://en.cppreference.com/w/cpp/container/deque).

Reply:
done